project3/participant/tasks/multicast_receiver_task.h

```diff
@@ -0,0 +1,52 @@
+/**
+ * @file Participant messaging task
+ */
+#ifndef PROJECT3_MULTICAST_RECEIVER_H
+#define PROJECT3_MULTICAST_RECEIVER_H
+
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <pub_sub_messages.pb.h>
+#include <sys/socket.h>
+#include <wire_protocol/wire_protocol.h>
+
+#include <cstdint>
+#include <memory>
+#include <string>
+
+#include "../participant_util.h"
+#include "console_task.h"
+#include "thread_pool/task.h"
+
+namespace participant_tasks {
+class MulticastReceiver : public thread_pool::Task {
+ public:
+  /**
+   * @param console_task the console_task that this receiver will print to
+   * @param log_location the location of the log file to log messages
+   * @param port the port this receiver will listen on
+   */
+  MulticastReceiver(const std::shared_ptr<ConsoleTask>& console_task,
+                    std::string log_location, int port);
+
+  Status SetUp() override;
+  Status RunAtomic() override;
+  void CleanUp() override;
+
+ private:
+  int messenger_fd_;
+  int port_;
+  std::string log_location_;
+  std::shared_ptr<ConsoleTask> console_task_;
+
+  /// buffer size for client.
+  static constexpr size_t kBufferSize = 4096;
+
+  /// buffer that stores serialized data to be received from the coordinator
+  std::vector<uint8_t> incoming_msg_buf_{};
+
+  /// parser for handling messages
+  wire_protocol::MessageParser<google::protobuf::Message> parser_;
```

Comment:
Shouldn't the template parameter here be `ForwardMulticast`?

Reply:
You're right, it can definitely be `ForwardMulticast`. I think I originally had other ideas for this parser but never adjusted it later.